integrators/ordinary_differential_equations.hpp

```diff
@@ -34,9 +34,137 @@ using base::Status;
+
+  using Flow = std::function<Status(Instant const& t_initial,
+                                    Instant const& t_final,
+                                    State const& initial_state,
+                                    State& final_state)>;
+
+  struct SystemState final {
+    SystemState() = default;
+    SystemState(State const& y, Instant const& t);
+
+    std::tuple<std::vector<DoublePrecision<StateTypes>>...> y;
+    DoublePrecision<Instant> time;
+
+    friend bool operator==(SystemState const& lhs, SystemState const& rhs) {
+      return lhs.y == rhs.y && lhs.time == rhs.time;
+    }
+  };
+
+  // We cannot use |Difference<StateTypes>| here for the same reason.  For some
+  // reason |DoublePrecision<StateTypes>| above works...
+  using SystemStateError =
+      std::tuple<std::vector<Difference<StateTypes, StateTypes>>...>;
+
+  // left_flow(t₀, t₁, X₀, X₁) sets X₁ to exp((t₁-t₀)A)X₀, and
+  // right_flow(t₀, t₁, X₀, X₁) sets X₁ to exp((t₁-t₀)B)X₀.
+  // The |std::vectors| in X₁ must have the same |size()| as those in X₀.  There
+  // is no other requirement on their values.
+  Flow left_flow;
+  Flow right_flow;
+};
+
+// A differential equation of the form q″ = f(q, q′, t).
+// |Position| is the type of q.
+template<typename Position_>
+struct ExplicitSecondOrderOrdinaryDifferentialEquation final {
+  using Position = Position_;
+  // The type of Δq.
+  using Displacement = Difference<Position>;
+  // The type of q′.
+  using Velocity = Variation<Position>;
+  // The type of q″.
+  using Acceleration = Variation<Velocity>;
+  using RightHandSideComputation =
+      std::function<Status(Instant const& t,
+                           std::vector<Position> const& positions,
+                           std::vector<Velocity> const& velocities,
+                           std::vector<Acceleration>& accelerations)>;
+
+  struct SystemState final {
+    SystemState() = default;
+    SystemState(std::vector<Position> const& q,
+                std::vector<Velocity> const& v,
+                Instant const& t);
+
+    std::vector<DoublePrecision<Position>> positions;
+    std::vector<DoublePrecision<Velocity>> velocities;
+    DoublePrecision<Instant> time;
+
+    friend bool operator==(SystemState const& lhs, SystemState const& rhs) {
+      return lhs.positions == rhs.positions &&
+             lhs.velocities == rhs.velocities &&
+             lhs.time == rhs.time;
+    }
+
+    void WriteToMessage(not_null<serialization::SystemState*> message) const;
+    static SystemState ReadFromMessage(
+        serialization::SystemState const& message);
+  };
+
+  struct SystemStateError final {
+    std::vector<Displacement> position_error;
+    std::vector<Velocity> velocity_error;
+  };
+
+  // A functor that computes f(q, t) and stores it in |accelerations|.
+  // This functor must be called with |accelerations.size()| equal to
+  // |positions.size()|, but there is no requirement on the values in
+  // |acceleration|.
```

Comment:
Comment the type, not the member.  Also, `accelerations`.

Reply:
Same as above.